src/main/ruby/truffleruby/core/splitter.rb

```diff
@@ -36,157 +36,225 @@
+
+        while match = Truffle::RegexpOperations.match(pattern, string, start)
+          break if limited && limit - count <= 1
+
+          collapsed = Truffle::RegexpOperations.collapsing?(match)
+
+          unless collapsed && (match.byte_begin(0) == last_match_end)
+            substring = Truffle::RegexpOperations.pre_match_from(match, last_match_end)
+            empty_count = add_substring(string, ret, substring, empty_count, &block)
+
+            # length > 1 means there are captures
+            if match.length > 1
+              match.captures.compact.each do |capture|
+                empty_count = add_substring(string, ret, capture, empty_count, &block)
+              end
+            end
+
+            count += 1
+          end
 
-      # Trim from end
-      if Primitive.undefined?(limit) || limit == 0
-        while s = ret.at(-1) and s.empty?
-          ret.pop
+          start = match.byte_end(0)
+          if collapsed
+            start += 1
+          end
+
+          last_match = match
+          last_match_end = last_match.byte_end(0)
         end
-      end
 
-      ret
-    end
+        if last_match
+          empty_count = add_substring(string, ret, last_match.post_match, empty_count, &block)
+        elsif ret.empty?
+          empty_count = add_substring(string, ret, string.dup, empty_count, &block)
+        end
+
+        if tail_empty || (!Primitive.undefined?(limit) && limit > 0)
+          add_empty(string, ret, empty_count, &block)
+        end
 
-    def self.split_on_string(string, pattern, trim_end)
-      pos = 0
+        result(string, ret, &block)
+      end
 
-      ret = []
 
-      pat_size = pattern.bytesize
-      str_size = string.bytesize
+      def add_substring(string, array, substring, empty_count, &block)
+        return empty_count + 1 if substring.length == 0 # remember another one empty match
 
-      while pos < str_size
-        nxt = Primitive.find_string(string, pattern, pos)
-        break unless nxt
+        add_empty(string, array, empty_count, &block)
 
-        match_size = nxt - pos
-        ret << string.byteslice(pos, match_size)
+        add_or_call(array, substring, &block)
 
-        pos = nxt + pat_size
+        0 # always release all empties
       end
 
-      # No more separators, but we need to grab the last part still.
-      ret << string.byteslice(pos, str_size - pos)
+      def add_empty(string, array, count, &block)
+        count.times { add_or_call(array, empty_string(string), &block) }
+      end
 
-      if trim_end
-        while s = ret.at(-1) and s.empty?
-          ret.pop
+      def add_or_call(array, element, &block)
+        if block_given?
+          block.call(element)
+        else
+          array << element
         end
       end
 
-      ret
+      def empty_string(original)
+        # Use #byteslice because it returns the right class and taints automatically.
+        original.byteslice(0,0)
+      end
+
+      def result(string, res, &block)
```

Comment:
Style wise, I think we usually lean towards inlining small helper methods.

Reply:
It just for DRY but I can inline it if necessary

Reply:
I'd tend to inline it too here here, and express it as a simple ternary condition `block ? string : res`.